File: modules/ls-semanticdb/src/ls/semanticdb/groups.scala
package ls.semanticdb

import ls.index.{NormalizedDocument, Role, SymbolKey, SymKind, SymProps, UnsafeReason}
import scala.collection.mutable

/** Exact alias groups for one ingest batch.
  *
  * `refGroups` drive `textDocument/references`; `renameGroups` drive
  * cross-file rename. In the v1 policy the two partitions are identical, but
  * they are separate fields because they are expected to diverge (rename must
  * stay conservative while references can merge more).
  *
  * Group vectors are deterministic: ordered by their minimal member under
  * (semanticSymbol, localDocId) ordering.
  */
final case class AliasGroups(
    refGroups: Vector[Set[SymbolKey]],
    refGroupIndex: Map[SymbolKey, Int],
    renameGroups: Vector[Set[SymbolKey]],
    renameGroupIndex: Map[SymbolKey, Int],
    /** Pure-semantic unsafe bits per rename group, computable without
      * document facts. Currently only [[UnsafeReason.OverrideFamily]];
      * doc-dependent bits (generated/readonly/external/dependency) are added
      * by [[RenameProfileBuilder]].
      */
    renameGroupSemanticMask: Vector[Long]
):
  def refGroupOf(key: SymbolKey): Option[Int] = refGroupIndex.get(key)
  def renameGroupOf(key: SymbolKey): Option[Int] = renameGroupIndex.get(key)

/** Builds EXACT alias groups by union-find over every SymbolKey seen in the
  * batch (symbols and occurrences).
  *
  * v1 merge policy (applied to BOTH ref and rename groups):
  *   - class/trait/enum `X#` merges with ALL its constructors
  *     `X#`<init>`(...)` (the class key is synthesized when only the
  *     constructor is present, e.g. a batch that only contains `new X`);
  *   - `X#` merges with its companion object `X.` when present in the batch;
  *   - companion `apply`/`unapply` methods (`X.apply(...).`,
  *     `X.unapply(...).`) merge into the class group when the class `X#` is
  *     present — their call-site token is the class name. Plain objects
  *     without a companion class do NOT merge with their apply methods;
  *   - a setter `x_=(...)` merges with its getter `x().` and with a val/var
  *     field term `x.` when those keys are present;
  *   - method overloads (same name, different disambiguator) stay in
  *     separate groups;
  *   - local symbols never merge.
  */
object AliasGroupBuilder:

  private val keyOrdering: Ordering[SymbolKey] =
    Ordering.by((k: SymbolKey) => (k.semanticSymbol, k.localDoc.fold(-1L)(_.value)))

  def build(docs: Vector[NormalizedDocument]): AliasGroups =
    // 1. Deterministic key universe + first-wins SymbolInfo lookup.
    val keySet = mutable.LinkedHashSet.empty[SymbolKey]
    val infoByKey = mutable.HashMap.empty[SymbolKey, ls.index.SymbolInfo]
    for doc <- docs do
      for s <- doc.symbols do
        keySet += s.key
        if !infoByKey.contains(s.key) then infoByKey.update(s.key, s)
      for o <- doc.occurrences do keySet += o.key

    // Constructors unambiguously imply their class symbol: synthesize the
    // class key so `new X` occurrences group with `X#` even when the class
    // is otherwise absent from this batch.
    val implied = mutable.LinkedHashSet.empty[SymbolKey]
    for key <- keySet if !key.isLocal do
      if SymbolStrings.isConstructor(key.semanticSymbol) then
        SymbolStrings.splitLast(key.semanticSymbol) match
          case Some((owner, _)) if owner.endsWith("#") =>
            implied += SymbolKey.global(owner)
          case Some(_) => ()
          case None => ()
    keySet ++= implied

    val keys = keySet.toVector.sorted(using keyOrdering)
    val indexOf: Map[SymbolKey, Int] = keys.iterator.zipWithIndex.toMap

    val uf = new UnionFind(keys.length)

    def unionWithGlobal(from: Int, toSymbol: String): Unit =
      indexOf.get(SymbolKey.global(toSymbol)) match
        case Some(to) => uf.union(from, to)
        case None => ()

    // 2. Apply the merge policy over global keys.
    for (key, i) <- keys.iterator.zipWithIndex if !key.isLocal do
      val sym = key.semanticSymbol
      SymbolStrings.splitLast(sym) match
        case Some((owner, SymbolStrings.Descriptor.Method(name, _))) =>
          if name == SymbolStrings.ConstructorName then
            if owner.endsWith("#") then unionWithGlobal(i, owner)
          else if name == "apply" || name == "unapply" then
            if owner.endsWith(".") then
              SymbolStrings.companion(owner) match
                case Some(cls) if indexOf.contains(SymbolKey.global(cls)) =>
                  unionWithGlobal(i, cls)
                case Some(_) => ()
                case None => ()
          else if name.length > 2 && name.endsWith("_=") then
            val base = SymbolStrings.encodeName(name.dropRight(2))
            unionWithGlobal(i, owner + base + "().")
            unionWithGlobal(i, owner + base + ".")
        case Some((owner, SymbolStrings.Descriptor.Type(name))) =>
          unionWithGlobal(i, owner + SymbolStrings.encodeName(name) + ".")
        case Some(_) => ()
        case None => ()

    // 2b. Export forwarders. Scala 3 `export B.m` synthesizes a forwarder
    //     method `A.m` in the exporting object: it carries a Method symbol but
    //     NO definition occurrence (no defining source token — the source shows
    //     `export B.m`), while the real `B.m` has one. So a Method symbol with
    //     no definition occurrence whose descriptor matches a definition-having
    //     Method under a different owner is a forwarder for that method. Union
    //     the forwarder into the original's group and remember it so the group
    //     is flagged UnsupportedSymbolFamily (exports are not safely renameable
    //     — a rename would have to touch the export clause and every forwarder).
    val definedKeys: Set[SymbolKey] =
      docs.iterator
        .flatMap(_.occurrences)
        .filter(_.role == Role.Definition)
        .map(_.key)
        .toSet
    def descriptorOf(sym: String): Option[(String, String)] =
      SymbolStrings.splitLast(sym).map((owner, _) => (owner, sym.substring(owner.length)))
    // definition-having methods indexed by their descriptor (name + signature)
    val definedMethodByDescriptor: Map[String, Vector[SymbolKey]] =
      keys.iterator
        .filter(k => !k.isLocal && definedKeys.contains(k))
        .filter(k => infoByKey.get(k).exists(_.kind == SymKind.Method))
        .flatMap(k => descriptorOf(k.semanticSymbol).map { case (owner, desc) => (desc, owner, k) })
        .toVector
        .groupMap(_._1)(t => t._3)
    val forwarderKeys = mutable.LinkedHashSet.empty[SymbolKey]
    for (key, i) <- keys.iterator.zipWithIndex if !key.isLocal do
      if infoByKey.get(key).exists(_.kind == SymKind.Method) && !definedKeys.contains(key) then
        descriptorOf(key.semanticSymbol) match
          case Some((owner, desc)) =>
            definedMethodByDescriptor.getOrElse(desc, Vector.empty).find { orig =>
              descriptorOf(orig.semanticSymbol).exists(_._1 != owner)
            } match
              case Some(orig) =>
                forwarderKeys += key
                uf.union(i, indexOf(orig))
              case None => ()
          case None => ()

    // 3. Assemble groups; iteration over sorted keys keeps output ordered by
    //    minimal member.
    val byRoot = mutable.LinkedHashMap.empty[Int, mutable.ArrayBuffer[SymbolKey]]
    for (key, i) <- keys.iterator.zipWithIndex do
      byRoot.getOrElseUpdate(uf.find(i), mutable.ArrayBuffer.empty[SymbolKey]) += key
    val groups: Vector[Set[SymbolKey]] = byRoot.valuesIterator.map(_.toSet).toVector
    val groupIndex: Map[SymbolKey, Int] =
      groups.iterator.zipWithIndex.flatMap((g, gi) => g.iterator.map(_ -> gi)).toMap

    // 4. OverrideFamily: a group is flagged when any member declares
    //    overridden symbols OR is itself overridden by another symbol in the
    //    batch (reverse map over overridden_symbols).
    val overriddenTargets: Set[SymbolKey] =
      infoByKey.valuesIterator
        .flatMap(_.overriddenSymbols.iterator)
        .map(SymbolKey.global)
        .toSet
    val semanticMask: Vector[Long] = groups.map { g =>
      var mask = 0L
      val overrideFlagged = g.exists { k =>
        infoByKey.get(k).exists(_.overriddenSymbols.nonEmpty) || overriddenTargets.contains(k)
      }
      if overrideFlagged then mask |= UnsafeReason.OverrideFamily
      if g.exists(forwarderKeys.contains) then mask |= UnsafeReason.UnsupportedSymbolFamily
      // Opaque types: rename is conservatively rejected (resolved policy). An
      // `opaque type T` merges with its companion and all uses, but renaming it
      // cannot be proven safe in v1, so the group is flagged unsafe outright.
      if g.exists(k => infoByKey.get(k).exists(i => (i.properties & SymProps.Opaque) != 0)) then
        mask |= UnsafeReason.OpaqueType
      mask
    }

    AliasGroups(
      refGroups = groups,
      refGroupIndex = groupIndex,
      renameGroups = groups,
      renameGroupIndex = groupIndex,
      renameGroupSemanticMask = semanticMask
    )

  /** Classic union-find with path compression and union by rank. */
  private final class UnionFind(size: Int):
    private val parent = Array.tabulate(size)(identity)
    private val rank = new Array[Int](size)

    def find(i: Int): Int =
      var root = i
      while parent(root) != root do root = parent(root)
      var cur = i
      while parent(cur) != root do
        val next = parent(cur)
        parent(cur) = root
        cur = next
      root

    def union(a: Int, b: Int): Unit =
      val ra = find(a)
      val rb = find(b)
      if ra != rb then
        if rank(ra) < rank(rb) then parent(ra) = rb
        else if rank(ra) > rank(rb) then parent(rb) = ra
        else
          parent(rb) = ra
          rank(ra) += 1

File: crates/ls-semanticdb/src/groups.rs
//! Exact alias groups for one ingest batch.
//!
//! `ref_groups` drive `textDocument/references`; `rename_groups` drive
//! cross-file rename. In the v1 policy the two partitions are identical, but they
//! are separate fields because they are expected to diverge (rename must stay
//! conservative while references can merge more). Group vectors are deterministic:
//! ordered by their minimal member under `(semantic_symbol, local_doc)` ordering.

use std::collections::{HashMap, HashSet};

use ls_index_model::{
    sym_props, unsafe_reason, NormalizedDocument, Role, SymKind, SymbolInfo, SymbolKey,
};

use crate::symbols::{self, Descriptor};

/// Exact alias groups plus the symbol -> group-index maps and the pure-semantic
/// unsafe mask per rename group.
#[derive(PartialEq, Eq)]
pub struct AliasGroups {
    pub ref_groups: Vec<HashSet<SymbolKey>>,
    pub ref_group_index: HashMap<SymbolKey, usize>,
    pub rename_groups: Vec<HashSet<SymbolKey>>,
    pub rename_group_index: HashMap<SymbolKey, usize>,
    /// Pure-semantic unsafe bits per rename group, computable without document
    /// facts (`OverrideFamily`, `UnsupportedSymbolFamily`, `OpaqueType`);
    /// doc-dependent bits are added by the rename-profile builder.
    pub rename_group_semantic_mask: Vec<u64>,
}

impl AliasGroups {
    pub fn ref_group_of(&self, key: &SymbolKey) -> Option<usize> {
        self.ref_group_index.get(key).copied()
    }
    pub fn rename_group_of(&self, key: &SymbolKey) -> Option<usize> {
        self.rename_group_index.get(key).copied()
    }
}

/// Sort key mirroring the Scala `(semanticSymbol, localDoc.fold(-1L)(_.value))`.
fn ord_val(k: &SymbolKey) -> i64 {
    k.local_doc.map(|d| d.value() as i64).unwrap_or(-1)
}

/// Builds EXACT alias groups by union-find over every `SymbolKey` seen in the
/// batch (symbols and occurrences), applying the v1 merge policy.
pub fn build(docs: &[NormalizedDocument]) -> AliasGroups {
    // 1. Key universe + first-wins SymbolInfo lookup.
    let mut key_set: HashSet<SymbolKey> = HashSet::new();
    let mut info_by_key: HashMap<SymbolKey, SymbolInfo> = HashMap::new();
    for doc in docs {
        for s in &doc.symbols {
            key_set.insert(s.key.clone());
            info_by_key
                .entry(s.key.clone())
                .or_insert_with(|| s.clone());
        }
        for o in &doc.occurrences {
            key_set.insert(o.key.clone());
        }
    }

    // Constructors imply their class symbol: synthesize the class key so `new X`
    // occurrences group with `X#` even when the class is otherwise absent.
    let mut implied: Vec<SymbolKey> = Vec::new();
    for key in &key_set {
        if !key.is_local() && symbols::is_constructor(&key.semantic_symbol) {
            if let Some((owner, _)) = symbols::split_last(&key.semantic_symbol) {
                if owner.ends_with('#') {
                    implied.push(SymbolKey::global(owner));
                }
            }
        }
    }
    key_set.extend(implied);

    let mut keys: Vec<SymbolKey> = key_set.into_iter().collect();
    keys.sort_by(|a, b| {
        a.semantic_symbol
            .cmp(&b.semantic_symbol)
            .then_with(|| ord_val(a).cmp(&ord_val(b)))
    });
    let index_of: HashMap<SymbolKey, usize> = keys
        .iter()
        .enumerate()
        .map(|(i, k)| (k.clone(), i))
        .collect();

    let mut uf = UnionFind::new(keys.len());

    let union_with_global = |uf: &mut UnionFind, from: usize, to_symbol: &str| {
        if let Some(&to) = index_of.get(&SymbolKey::global(to_symbol)) {
            uf.union(from, to);
        }
    };

    // 2. Merge policy over global keys.
    for (i, key) in keys.iter().enumerate() {
        if key.is_local() {
            continue;
        }
        let sym = &key.semantic_symbol;
        match symbols::split_last(sym) {
            Some((owner, Descriptor::Method(name, _))) => {
                if name == symbols::CONSTRUCTOR_NAME {
                    if owner.ends_with('#') {
                        union_with_global(&mut uf, i, &owner);
                    }
                } else if name == "apply" || name == "unapply" {
                    if owner.ends_with('.') {
                        if let Some(cls) = symbols::companion(&owner) {
                            if index_of.contains_key(&SymbolKey::global(&cls)) {
                                union_with_global(&mut uf, i, &cls);
                            }
                        }
                    }
                } else if name.chars().count() > 2 && name.ends_with("_=") {
                    let mut base_src = name.clone();
                    base_src.truncate(base_src.len() - 2);
                    let base = symbols::encode_name(&base_src);
                    union_with_global(&mut uf, i, &format!("{owner}{base}()."));
                    union_with_global(&mut uf, i, &format!("{owner}{base}."));
                }
            }
            Some((owner, Descriptor::Type(name))) => {
                union_with_global(
                    &mut uf,
                    i,
                    &format!("{owner}{}.", symbols::encode_name(&name)),
                );
            }
            _ => {}
        }
    }

    // 2b. Export forwarders. A Scala 3 `export B.m` synthesizes a forwarder
    // method `A.m` with NO definition occurrence, while the real `B.m` has one.
    // Union a definition-less Method into the UNIQUE definition-having method
    // that shares its descriptor under a different owner, and flag the group.
    let defined_keys: HashSet<SymbolKey> = docs
        .iter()
        .flat_map(|d| d.occurrences.iter())
        .filter(|o| o.role == Role::Definition)
        .map(|o| o.key.clone())
        .collect();

    // definition-having methods indexed by their descriptor (name + signature)
    let mut defined_method_by_descriptor: HashMap<String, Vec<SymbolKey>> = HashMap::new();
    for k in &keys {
        if k.is_local() || !defined_keys.contains(k) {
            continue;
        }
        if info_by_key.get(k).map(|i| i.kind) != Some(SymKind::Method) {
            continue;
        }
        if let Some((_, desc)) = descriptor_split(&k.semantic_symbol) {
            defined_method_by_descriptor
                .entry(desc)
                .or_default()
                .push(k.clone());
        }
    }

    let mut forwarder_keys: HashSet<SymbolKey> = HashSet::new();
    for (i, key) in keys.iter().enumerate() {
        if key.is_local() {
            continue;
        }
        let is_method = info_by_key.get(key).map(|i| i.kind) == Some(SymKind::Method);
        if !is_method || defined_keys.contains(key) {
            continue;
        }
        if let Some((owner, desc)) = descriptor_split(&key.semantic_symbol) {
            // Union ONLY when the descriptor maps to a UNIQUE definition-having
            // method under a different owner; skip on ambiguous or absent match.
            let mut candidates: Vec<SymbolKey> = defined_method_by_descriptor
                .get(&desc)
                .map(|v| v.as_slice())
                .unwrap_or(&[])
                .iter()
                .filter(|orig| {
                    descriptor_split(&orig.semantic_symbol)
                        .map(|(o, _)| o != owner)
                        .unwrap_or(false)
                })
                .cloned()
                .collect();
            candidates.dedup();
            if let [orig] = candidates.as_slice() {
                forwarder_keys.insert(key.clone());
                uf.union(i, index_of[orig]);
            }
        }
    }

    // 3. Assemble groups; iteration over sorted keys keeps output ordered by
    //    minimal member.
    let mut order: Vec<usize> = Vec::new();
    let mut buckets: HashMap<usize, Vec<SymbolKey>> = HashMap::new();
    for (i, key) in keys.iter().enumerate() {
        let root = uf.find(i);
        if !buckets.contains_key(&root) {
            order.push(root);
        }
        buckets.entry(root).or_default().push(key.clone());
    }
    let groups: Vec<HashSet<SymbolKey>> = order
        .iter()
        .map(|r| buckets[r].iter().cloned().collect())
        .collect();
    let mut group_index: HashMap<SymbolKey, usize> = HashMap::new();
    for (gi, g) in groups.iter().enumerate() {
        for k in g {
            group_index.insert(k.clone(), gi);
        }
    }

    // 4. OverrideFamily / UnsupportedSymbolFamily / OpaqueType semantic mask.
    let overridden_targets: HashSet<SymbolKey> = info_by_key
        .values()
        .flat_map(|i| i.overridden_symbols.iter())
        .map(SymbolKey::global)
        .collect();
    let semantic_mask: Vec<u64> = groups
        .iter()
        .map(|g| {
            let mut mask = 0u64;
            let override_flagged = g.iter().any(|k| {
                info_by_key
                    .get(k)
                    .map(|i| !i.overridden_symbols.is_empty())
                    .unwrap_or(false)
                    || overridden_targets.contains(k)
            });
            if override_flagged {
                mask |= unsafe_reason::OVERRIDE_FAMILY;
            }
            if g.iter().any(|k| forwarder_keys.contains(k)) {
                mask |= unsafe_reason::UNSUPPORTED_SYMBOL_FAMILY;
            }
            if g.iter().any(|k| {
                info_by_key
                    .get(k)
                    .map(|i| i.properties & sym_props::OPAQUE != 0)
                    .unwrap_or(false)
            }) {
                mask |= unsafe_reason::OPAQUE_TYPE;
            }
            mask
        })
        .collect();

    AliasGroups {
        ref_groups: groups.clone(),
        ref_group_index: group_index.clone(),
        rename_groups: groups,
        rename_group_index: group_index,
        rename_group_semantic_mask: semantic_mask,
    }
}

/// `splitLast(sym).map((owner, _) => (owner, sym.substring(owner.length)))` —
/// the owner prefix plus the raw descriptor suffix (name + signature).
fn descriptor_split(sym: &str) -> Option<(String, String)> {
    symbols::split_last(sym).map(|(owner, _)| {
        let owner_chars = owner.chars().count();
        let desc: String = sym.chars().skip(owner_chars).collect();
        (owner, desc)
    })
}

/// Classic union-find with path compression and union by rank.
struct UnionFind {
    parent: Vec<usize>,
    rank: Vec<u32>,
}

impl UnionFind {
    fn new(size: usize) -> Self {
        UnionFind {
            parent: (0..size).collect(),
            rank: vec![0; size],
        }
    }

    fn find(&mut self, i: usize) -> usize {
        let mut root = i;
        while self.parent[root] != root {
            root = self.parent[root];
        }
        let mut cur = i;
        while self.parent[cur] != root {
            let next = self.parent[cur];
            self.parent[cur] = root;
            cur = next;
        }
        root
    }

    fn union(&mut self, a: usize, b: usize) {
        let ra = self.find(a);
        let rb = self.find(b);
        if ra == rb {
            return;
        }
        match self.rank[ra].cmp(&self.rank[rb]) {
            std::cmp::Ordering::Less => self.parent[ra] = rb,
            std::cmp::Ordering::Greater => self.parent[rb] = ra,
            std::cmp::Ordering::Equal => {
                self.parent[rb] = ra;
                self.rank[ra] += 1;
            }
        }
    }
}
